### ISSUE TITLE
[Windows][Desktop]: WidgetsBindingObserver is not working.

### ISSUE THREAD

**Issue by WaadSulaiman** [flutter/flutter#98355]
WidgetsBindingObserver

Doesn't seem to work on flutter desktop. When switching the application to run on web. It's working fine.

Sincerely!

**Comment by darshankawar** [comment]
@WaadSulaiman 
Can you provide `flutter doctor -v` and a [minimal reproducible code sample](https://stackoverflow.com/help/minimal-reproducible-example) that shows the behavior you are reporting along with expected result you are looking for ?

**Comment by WaadSulaiman** [comment]
@darshankawar 

The application is working fine.

Here is ```flutter doctor -v ``` output:
```
[√] Flutter (Channel stable, 2.10.0, on Microsoft Windows [Version 10.0.22000.493], locale en-SE)
    • Flutter version 2.10.0 at C:\Applications\flutter
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision 5f105a6ca7 (13 days ago), 2022-02-01 14:15:42 -0800
    • Engine revision 776efd2034
    • Dart version 2.16.0
    • DevTools version 2.9.2

[√] Android toolchain - develop for Android devices (Android SDK version 30.0.3)
    • Android SDK at C:\Users\waads\AppData\Local\Android\sdk
    • Platform android-31, build-tools 30.0.3
    • Java binary at: C:\Program Files\Android\Android Studio\jre\bin\java
    • Java version OpenJDK Runtime Environment (build 11.0.11+9-b60-7590822)
    • All Android licenses accepted.

[√] Chrome - develop for the web
    • Chrome at C:\Program Files\Google\Chrome\Application\chrome.exe

[√] Visual Studio - develop for Windows (Visual Studio Community 2022 17.0.6)
    • Visual Studio at C:\Program Files\Microsoft Visual Studio\2022\Community
    • Visual Studio Community 2022 version 17.0.32126.317
    • Windows 10 SDK version 10.0.19041.0

[√] Android Studio (version 2021.1)
    • Android Studio at C:\Program Files\Android\Android Studio
    • Flutter plugin can be installed from:
       https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
       https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 11.0.11+9-b60-7590822)

[√] VS Code (version 1.64.2)
    • VS Code at C:\Users\waads\AppData\Local\Programs\Microsoft VS Code
    • Flutter extension version 3.34.0

[√] Connected device (3 available)
    • Windows (desktop) • windows • windows-x64    • Microsoft Windows [Version 10.0.22000.493]
    • Chrome (web)      • chrome  • web-javascript • Google Chrome 98.0.4758.82
    • Edge (web)        • edge    • web-javascript • Microsoft Edge 98.0.1108.50

[√] HTTP Host Availability
    • All required HTTP hosts are available

• No issues found!
```
The problem is that the functions I'm overriding are not getting triggered.
Here is the code:
```dart
class MyApp extends StatefulWidget {
  const MyApp({Key? key}) : super(key: key);

  @override
  _MyAppState createState() => _MyAppState();
}

class _MyAppState extends State<MyApp> with WidgetsBindingObserver {
  
  @override
  void initState() {
    super.initState();
    WidgetsBinding.instance!.addObserver(this);
  }

  @override
  void dispose() {
    WidgetsBinding.instance!.removeObserver(this);
    super.dispose();
  }

  @override
  void didChangePlatformBrightness() {
    super.didChangePlatformBrightness();
    context.read<ThemeProvider>().setThemeData(window.platformBrightness);
  }

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      home: const HomePage(),
    );
```
```didChangePlatformBrightness``` is getting triggered when running on Web, Android, iOS. But when running it on desktop, such as Windows. The function is not getting triggered.

Sincerely!

**Comment by darshankawar** [comment]
Thanks for the details. I'll look into it more. Mean time, can you provide missing code for `context.read<ThemeProvider>().setThemeData(window.platformBrightness);`

**Comment by WaadSulaiman** [comment]
@darshankawar 

Here is the function for ```context.read<ThemeProvider>().setThemeData(window.platformBrightness);```

```dart
  void setThemeData(Brightness brightness) {
        if (brightness == Brightness.light) {
              _themeData = lightThemeData;
        } else {
              _themeData = darkThemeData;
        }
        notifyListeners();
  }
```

Once again the code is correct since it's working on Android, IOS and Web.

Sincerely!

**Comment by darshankawar** [comment]
@WaadSulaiman 
This is the complete code sample I came up with using the one you provided.

<details>
<summary> complete sample </summary>

```
import 'dart:ui';

import 'package:flutter/material.dart';

class MyApp extends StatefulWidget {
  const MyApp({Key? key}) : super(key: key);

  @override
  _MyAppState createState() => _MyAppState();
}

class _MyAppState extends State<MyApp> with WidgetsBindingObserver {

  @override
  void initState() {
    super.initState();
    WidgetsBinding.instance!.addObserver(this);
  }

  @override
  void dispose() {
    WidgetsBinding.instance!.removeObserver(this);
    super.dispose();
  }

  @override
  void didChangePlatformBrightness() {
    super.didChangePlatformBrightness();
    context.read<ThemeProvider>().setThemeData(window.platformBrightness);
  }

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      home: const MyApp(),
    );
  }
}

class ThemeProvider {
}
void setThemeData(Brightness brightness) {
  if (brightness == Brightness.light) {
    _themeData = lightThemeData;
  } else {
    _themeData = darkThemeData;
  }
  notifyListeners();
}

```

</details>

I am still getting compile errors on `ThemeProvider` class above.

**Comment by jayjah** [comment]
@darshankawar I think you should put the following method
```
void setThemeData(Brightness brightness) {
  if (brightness == Brightness.light) {
    _themeData = lightThemeData;
  } else {
    _themeData = darkThemeData;
  }
  notifyListeners();
}
```
INSIDE the class of `ThemeProvider`. 
Just like below:
```
class ThemeProvider {
void setThemeData(Brightness brightness) {
  if (brightness == Brightness.light) {
    _themeData = lightThemeData;
  } else {
    _themeData = darkThemeData;
  }
  notifyListeners();
}
}
```

The code you provided does the opposite and puts the method outside of the class.

**Comment by WaadSulaiman** [comment]
Hey, 

Mate could you please just label it as a bug. Once again the code is working fine on Android, IOS, Web. It's not working on desktop. About your sample, it should look like the code @jayjah provided. 

Sincerely!

**Comment by darshankawar** [comment]
Hey @WaadSulaiman ,
I can't just label it as a bug without verifying at our end. Also, the code you provided was not complete when we requested you to give us complete minimal code sample. I was getting compile error on `context.read<ThemeProvider>()`, specially on `read`. I had to research it and then found out that it comes from `provider` package and then had to add it, which you didn't specify when you provided the code. Please don't assume that everybody will know what package one has to add in order to make the code sample work.

Also, below is the complete code sample which is now compile-free, but running it is giving me below errors in console:

<details>
<summary> complete code sample </summary>

```
import 'dart:ui';

import 'package:flutter/material.dart';
import 'package:provider/provider.dart';

ThemeData _themeData = ThemeData(
  brightness: Brightness.light
);

ThemeData _lightThemeData = ThemeData(
  brightness: Brightness.dark
);

void main() {
  runApp(MaterialApp(
    home: MyApp(),
  ));
}

class MyApp extends StatefulWidget {
  const MyApp({Key? key}) : super(key: key);

  @override
  _MyAppState createState() => _MyAppState();
}

class _MyAppState extends State<MyApp> with WidgetsBindingObserver {

  @override
  void initState() {
    super.initState();
    WidgetsBinding.instance.addObserver(this);
  }

  @override
  void dispose() {
    WidgetsBinding.instance.removeObserver(this);
    super.dispose();
  }

  @override
  void didChangePlatformBrightness() {
    super.didChangePlatformBrightness();
    context.read<ThemeProvider>().setThemeData(window.platformBrightness);
  }

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      home: const MyApp(),
    );
  }
}

class ThemeProvider with ChangeNotifier{
  void setThemeData(Brightness brightness) {
    if (brightness == Brightness.light) {
      _themeData = _themeData;
    } else {
      _themeData = _lightThemeData;
    }
    notifyListeners();
  }
}


```

</details>


<details>
<summary> console error log </summary>

```

══╡ EXCEPTION CAUGHT BY WIDGETS LIBRARY ╞═══════════════════════════════════════════════════════════
The following StackOverflowError was thrown building _FocusMarker:
Stack Overflow

The relevant error-causing widget was:
  MaterialApp MaterialApp:file:///Users/dhs/Documents/NCFlutter/testdemo/lib/main.dart:49:12

When the exception was thrown, this was the stack:
#0      int._toPow2String (dart:core-patch/integers.dart:365:3)
#1      int.toRadixString (dart:core-patch/integers.dart:339:14)
#2      shortHash (package:flutter/src/foundation/diagnostics.dart:3047:41)
#3      describeIdentity (package:flutter/src/foundation/diagnostics.dart:3059:95)
#4      Animation.toString (package:flutter/src/animation/animation.dart:175:15)
#5      _StringBase._interpolate (dart:core-patch/string_patch.dart:853:19)
#6      ProxyAnimation.toString (package:flutter/src/animation/animations.dart:238:71)
#7      _StringBase._interpolateSingle (dart:core-patch/string_patch.dart:831:17)
#8      StringBuffer.write (dart:core-patch/string_buffer_patch.dart:64:24)
#9      List._joinWithSeparator (dart:core-patch/growable_array.dart:497:12)
#10     List.join (dart:core-patch/growable_array.dart:442:38)
#11     _MergingListenable.toString (package:flutter/src/foundation/change_notifier.dart:401:43)
#12     DiagnosticsProperty.valueToString (package:flutter/src/foundation/diagnostics.dart:2777:60)
#13     DiagnosticsProperty.toDescription (package:flutter/src/foundation/diagnostics.dart:2791:21)
#14     TextTreeRenderer._debugRender (package:flutter/src/foundation/diagnostics.dart:1189:32)
#15     TextTreeRenderer.render (package:flutter/src/foundation/diagnostics.dart:1121:14)
#16     TextTreeRenderer._debugRender (package:flutter/src/foundation/diagnostics.dart:1299:39)
#17     TextTreeRenderer.render (package:flutter/src/foundation/diagnostics.dart:1121:14)
#18     DiagnosticsNode.toStringDeep.<anonymous closure> (package:flutter/src/foundation/diagnostics.dart:1776:9)
#19     DiagnosticsNode.toStringDeep (package:flutter/src/foundation/diagnostics.dart:1783:6)
#20     DiagnosticsNode.toString.<anonymous closure> (package:flutter/src/foundation/diagnostics.dart:1694:18)
#21     DiagnosticsNode.toString (package:flutter/src/foundation/diagnostics.dart:1706:6)
#22     Diagnosticable.toString.<anonymous closure> (package:flutter/src/foundation/diagnostics.dart:3168:78)
#23     Diagnosticable.toString (package:flutter/src/foundation/diagnostics.dart:3170:6)
#24     _StringBase._interpolate (dart:core-patch/string_patch.dart:853:19)
#25     ComponentElement.performRebuild (package:flutter/src/widgets/framework.dart:4802:44)
#26     StatefulElement.performRebuild (package:flutter/src/widgets/framework.dart:4963:11)
#27     Element.rebuild (package:flutter/src/widgets/framework.dart:4515:5)
#28     ComponentElement._firstBuild (package:flutter/src/widgets/framework.dart:4773:5)
#29     StatefulElement._firstBuild (package:flutter/src/widgets/framework.dart:4954:11)
#30     ComponentElement.mount (package:flutter/src/widgets/framework.dart:4767:5)
...     Normal element mounting (13 frames)
#43     Element.inflateWidget (package:flutter/src/widgets/framework.dart:3823:14)
...
...
#4569   ComponentElement.mount (package:flutter/src/widgets/framework.dart:4767:5)
...     Normal element mounting (21 frames)
#4590   Element.inflateWidget (package:flutter/src/widgets/framework.dart:3823:14)
#4591   Element.updateChild (package:flutter/src/widgets/framework.dart:3562:18)
#4592   RenderObjectToWidgetElement._rebuild (package:flutter/src/widgets/binding.dart:1193:16)
#4593   RenderObjectToWidgetElement.mount (package:flutter/src/widgets/binding.dart:1162:5)
#4594   RenderObjectToWidgetAdapter.attachToRenderTree.<anonymous closure> (package:flutter/src/widgets/binding.dart:1110:18)
#4595   BuildOwner.buildScope (package:flutter/src/widgets/framework.dart:2654:19)
#4596   RenderObjectToWidgetAdapter.attachToRenderTree (package:flutter/src/widgets/binding.dart:1109:13)
#4597   WidgetsBinding.attachRootWidget (package:flutter/src/widgets/binding.dart:944:7)
#4598   WidgetsBinding.scheduleAttachRootWidget.<anonymous closure> (package:flutter/src/widgets/binding.dart:924:7)
(elided 11 frames from class _RawReceivePortImpl, class _Timer, dart:async, and dart:async-patch)

════════════════════════════════════════════════════════════════════════════════════════════════════

Another exception was thrown: 'package:flutter/src/widgets/framework.dart': Failed assertion: line 4517 pos 14: 'owner!._debugCurrentBuildTarget == this': is not true.

Another exception was thrown: 'package:flutter/src/widgets/framework.dart': Failed assertion: line 4517 pos 14: 'owner!._debugCurrentBuildTarget == this': is not true.

Another exception was thrown: 'package:flutter/src/widgets/framework.dart': Failed assertion: line 4517 pos 14: 'owner!._debugCurrentBuildTarget == this': is not true.

Another exception was thrown: 'package:flutter/src/widgets/framework.dart': Failed assertion: line 4517 pos 14: 'owner!._debugCurrentBuildTarget == this': is not true.

Another exception was thrown: 'package:flutter/src/widgets/framework.dart': Failed assertion: line 4517 pos 14: 'owner!._debugCurrentBuildTarget == this': is not true.

Another exception was thrown: 'package:flutter/src/widgets/framework.dart': Failed assertion: line 4517 pos 14: 'owner!._debugCurrentBuildTarget == this': is not true.

Another exception was thrown: 'package:flutter/src/widgets/framework.dart': Failed assertion: line 4517 pos 14: 'owner!._debugCurrentBuildTarget == this': is not true.

Another exception was thrown: 'package:flutter/src/widgets/framework.dart': Failed assertion: line 4517 pos 14: 'owner!._debugCurrentBuildTarget == this': is not true.

Another exception was thrown: 'package:flutter/src/widgets/framework.dart': Failed assertion: line 4517 pos 14: 'owner!._debugCurrentBuildTarget == this': is not true.

Another exception was thrown: 'package:flutter/src/widgets/framework.dart': Failed assertion: line 4517 pos 14: 'owner!._debugCurrentBuildTarget == this': is not true.

Another exception was thrown: 'package:flutter/src/widgets/framework.dart': Failed assertion: line 4517 pos 14: 'owner!._debugCurrentBuildTarget == this': is not true.

Another exception was thrown: 'package:flutter/src/widgets/framework.dart': Failed assertion: line 4517 pos 14: 'owner!._debugCurrentBuildTarget == this': is not true.

Another exception was thrown: 'package:flutter/src/widgets/framework.dart': Failed assertion: line 4517 pos 14: 'owner!._debugCurrentBuildTarget == this': is not true.

Another exception was thrown: 'package:flutter/src/widgets/framework.dart': Failed assertion: line 4517 pos 14: 'owner!._debugCurrentBuildTarget == this': is not true.

Another exception was thrown: 'package:flutter/src/widgets/framework.dart': Failed assertion: line 4517 pos 14: 'owner!._debugCurrentBuildTarget == this': is not true.

Another exception was thrown: 'package:flutter/src/widgets/framework.dart': Failed assertion: line 4517 pos 14: 'owner!._debugCurrentBuildTarget == this': is not true.

Another exception was thrown: 'package:flutter/src/widgets/framework.dart': Failed assertion: line 4517 pos 14: 'owner!._debugCurrentBuildTarget == this': is not true.

Another exception was thrown: 'package:flutter/src/widgets/framework.dart': Failed assertion: line 4517 pos 14: 'owner!._debugCurrentBuildTarget == this': is not true.

```

</details>

So in order to avoid going back and forth on these, please provide a runnable code sample.

Also, check this similar issue and see if it's related to your case: https://github.com/flutter/flutter/issues/54612#

**Comment by WaadSulaiman** [comment]
Hey @darshankawar,

Here is the sample code.

```main.dart```
```dart
void main() {
  runApp(
    MultiProvider(
      providers: [
        ChangeNotifierProvider(
          create: (_) => ThemeProvider(),
        ),
      ],
      child: const MyApp(),
    ),
  );
}
```
```my_app.dart```

```dart
class MyApp extends StatefulWidget {
  const MyApp({Key? key}) : super(key: key);

  @override
  _MyAppState createState() => _MyAppState();
}

class _MyAppState extends State<MyApp> with WidgetsBindingObserver {
  @override
  void initState() {
    super.initState();
    WidgetsBinding.instance!.addObserver(this);
  }

  @override
  void dispose() {
    WidgetsBinding.instance!.removeObserver(this);
    super.dispose();
  }

  @override
  void didChangePlatformBrightness() {
    super.didChangePlatformBrightness();
    context.read<ThemeProvider>().setThemeData(window.platformBrightness);
  }

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      theme: context.watch<ThemeProvider>().themeData,
      home: const Scaffold(),
    );
  }
}
```
```theme_provider.dart```

```dart
class ThemeProvider with ChangeNotifier {
  ThemeProvider() {
    _themeData = ThemeData.light();
  }
  // Represents the current theme data.
  late ThemeData _themeData;
  // A getter for _themeData.
  ThemeData get themeData => _themeData;
  // for changing theme data.
  void setThemeData(Brightness brightness) {
    if (brightness == Brightness.light) {
      _themeData = ThemeData.light();
    } else {
      _themeData = ThemeData.dark();
    }
    notifyListeners();
  }
}
```
```pubspec.yaml```

```dart
dependencies:
  cupertino_icons: ^1.0.2
  flutter:
    sdk: flutter
  provider: ^6.0.2
```

**Comment by WaadSulaiman** [comment]
@darshankawar
Run the application on web and go into your computer settings and change the theme from light to dark or vice versa. ```WidgetsBindingObserver``` will catch the action and change the theme of the application. Which is correct, now do the same thing but this time run the application on desktop and you will notice that nothing happens. Which is why I opened an issue with the title  ```WidgetsBindingObserver not working on flutter desktop #98355```. This is what the conversation has been about, ```provider``` as a 3rd party plugin has nothing to do with this issue.

Sincerely!

**Comment by darshankawar** [comment]
Thanks for the update and patience. I verified this behavior on master and stable versions and observed that this is specific to windows desktop. ie, running the code sample provided and then trying to change the machine's theme from dark to light and vice-versa doesn't affect the Flutter desktop app and it remains same:

![win_d](https://user-images.githubusercontent.com/67046386/154424458-78bb6395-7c19-4b27-a052-1e086fad4c92.png)
![win_theme](https://user-images.githubusercontent.com/67046386/154424466-78b0322c-aebc-43bc-bfca-beb6e15f4911.png)

<details>
<summary> stable, master flutter doctor -v </summary>

```
[√] Flutter (Channel stable, 2.8.1, on Microsoft Windows [Version 10.0.19042.1466], locale en-US)
    • Flutter version 2.8.1 at D:\FlutterSDK\flutter
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision 77d935af4d (9 weeks ago), 2021-12-16 08:37:33 -0800
    • Engine revision 890a5fca2e
    • Dart version 2.15.1

[√] Android toolchain - develop for Android devices (Android SDK version 31.0.0)
    • Android SDK at C:\Users\Win\AppData\Local\Android\sdk
    • Platform android-31, build-tools 31.0.0
    • Java binary at: C:\Program Files\Android\Android Studio\jre\bin\java
    • Java version OpenJDK Runtime Environment (build 11.0.10+0-b96-7249189)
    • All Android licenses accepted.

[√] Chrome - develop for the web
    • Chrome at C:\Program Files (x86)\Google\Chrome\Application\chrome.exe

[√] Visual Studio - develop for Windows (Visual Studio Community 2019 16.11.2)
    • Visual Studio at C:\Program Files (x86)\Microsoft Visual Studio\2019\Community
    • Visual Studio Community 2019 version 16.11.31624.102
    • Windows 10 SDK version 10.0.19041.0

[√] Android Studio (version 2020.3)
    • Android Studio at C:\Program Files\Android\Android Studio
    • Flutter plugin can be installed from:
       https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
       https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 11.0.10+0-b96-7249189)

[√] VS Code (version 1.62.3)
    • VS Code at C:\Users\Win\AppData\Local\Programs\Microsoft VS Code
    • Flutter extension version 3.28.0

[√] Connected device (5 available)
    • SM A260G (mobile) • 5200763ebcfa861f • android-arm    • Android 8.1.0 (API 27)
    • SM G975F (mobile) • RZ8M802WY0X      • android-arm64  • Android 11 (API 30)
    • Windows (desktop) • windows          • windows-x64    • Microsoft Windows [Version 10.0.19042.1466]
    • Chrome (web)      • chrome           • web-javascript • Google Chrome 96.0.4664.110
    • Edge (web)        • edge             • web-javascript • Microsoft Edge 98.0.1108.50

• No issues found!


[√] Flutter (Channel master, 2.11.0-0.0.pre.542, on Microsoft Windows [Version 10.0.19042.1466], locale en-US)
[√] Android toolchain - develop for Android devices (Android SDK version 31.0.0)
[√] Chrome - develop for the web
[√] Visual Studio - develop for Windows (Visual Studio Community 2019 16.11.2)
[√] Android Studio (version 2020.3)
[√] VS Code (version 1.62.3)
[√] Connected device (5 available)
[√] HTTP Host Availability

• No issues found!

```
</details>

Also verified this behavior on macOS, wherein, the desktop app is updated per the theme selected, as below:

**Comment by WaadSulaiman** [comment]
Hey, @darshankawar

I changed the title since ```WidgetsBindingObserver``` is just there to provide us some additional functions. It has nothing to do with theme changing directly. It just captures events.

**Comment by gspencergoog** [comment]
I think this is because the system preferences dark mode detection isn't implemented yet on Windows. (Desktop triage).

**Comment by moko256** [comment]
Hello, I implemented theme change detection: flutter/engine#29303
